### PR TITLE
Add a build info metric to Prometheus

### PR DESCRIPTION
**Pull request by SuperQ**
Include `unifi_poller_build_info` as a Prometheus metric.

Signed-off-by: Ben Kochie <superq@gmail.com>

### CODE REVIEW

**Comment by SuperQ** [comment]
Example output:
```
# HELP unifi_poller_build_info A metric with a constant '1' value labeled by version, revision, branch, and goversion from which unifi_poller was built.
# TYPE unifi_poller_build_info gauge
unifi_poller_build_info{branch="superq/version_metric",goversion="go1.13.5",revision="4e79b0b",version="1.6.3-642"} 1
```

File: Makefile [hunk 1]
```diff
@@ -42,6 +42,8 @@ $(PACKAGE_SCRIPTS) \
 --config-files "/etc/$(BINARY)/$(CONFIG_FILE)"
 endef
 
+VERSION_LDFLAGS:=-X $(POLLER_PATH).Branch=$(BRANCH) -X $(POLLER_PATH).BuildDate=$(DATE) -X $(POLLER_PATH).Revision=$(COMMIT) -X $(POLLER_PATH).Version=$(VERSION)-$(ITERATION)
```
Comment (davidnewhall) [inline Makefile]: Couldn't we just do `-X github.com/prometheus/common/version.Branch=$(BRANCH)` and repeat for the others? Then we don't need new/more global variables, and we could in fact remove the `Version` variable by doing it this way.  Then there's only two new lines of Go code, the rest is in the Makefile. Thoughts?

File: pkg/poller/prometheus.go [hunk 2]
```diff
@@ -24,6 +25,12 @@ func (u *UnifiPoller) RunPrometheus() error {
 		ReportErrors: true, // XXX: Does this need to be configurable?
 	}))
 
+	version.Branch = Branch
+	version.BuildDate = BuildDate
+	version.Revision = Revision
+	version.Version = Version
+	prometheus.MustRegister(version.NewCollector("unifi_poller"))
```
Comment (davidnewhall) [inline pkg/poller/prometheus.go]: I really wish you could name it with a dash here. this bugs me haha

File: Makefile (hunk 1 above)
Comment (davidnewhall) [reply Makefile]: Also, please add a `\` or two to wrap this long line.

Comment (SuperQ) [reply Makefile]: We could go directly to the version package variable. But sometimes it's nicer to have it in the local vars for other uses like printing it out in the logs. I just didn't add that feature here.

File: pkg/poller/prometheus.go (hunk 2 above)
Comment (SuperQ) [reply pkg/poller/prometheus.go]: Dashes are not allowed in the metric name. :)

File: Makefile (hunk 1 above)
Comment (SuperQ) [reply Makefile]: So, when I try and use `-X github.com/prometheus/common/version...`, it doesn't seem to populate correctly. I think I'd rather stick to the current method of getting these variables set.

But I split up the long line.

Comment (davidnewhall) [reply Makefile]: This worked, but I don't know how portable it is if we move off of `dep` and into the module world.
```
-X github.com/davidnewhall/unifi-poller/vendor/github.com/prometheus/common/version.Version=$(VERSION)-$(ITERATION)
```

Comment (SuperQ) [reply Makefile]: Ahh, right! That jogged my memory from our setup before our go mod changes in Prometheus land.

With go mod, it's simplifies to `-X github.com/prometheus/common/version.Version`

**Comment by SuperQ** [comment]
Ok, cleaned up, PTAL

File: Makefile [hunk 3]
```diff
@@ -42,6 +42,12 @@ $(PACKAGE_SCRIPTS) \
 --config-files "/etc/$(BINARY)/$(CONFIG_FILE)"
 endef
 
+VERSION_LDFLAGS:= \
+  -X $(IMPORT_PATH)/vendor/github.com/prometheus/common/version.Branch=$(BRANCH) \
+  -X $(IMPORT_PATH)/vendor/github.com/prometheus/common/version.BuildDate=$(DATE) \
+  -X $(IMPORT_PATH)/vendor/github.com/prometheus/common/version.Revision=$(COMMIT) \
```
Comment (davidnewhall) [inline Makefile]: Does this seem OK though? You mentioned you'd rather not do this. I'm on the fence. I like the idea of containerizing the pieces in one package we can import from anywhere (ie, if want to expose these variables to the influxunifi or promunifi packages, it works). I don't like that `vendor/` is in this path though, that bugs me. I haven't looked into go modules yet, but I feel like this wont work if someone uses them. Maybe I shouldn't care about that and cross the bridge when we get there. What's your feeling on this?

File: Makefile [hunk 4]
```diff
@@ -42,6 +42,12 @@ $(PACKAGE_SCRIPTS) \
...
+  -X $(VERSION_PATH)=$(VERSION)-$(ITERATION)
```
Comment (davidnewhall) [inline Makefile]: I'd be fine converting this too, remove the local global variable and use `version.Version` directly. Pending your "thought" on my other comment.

File: Makefile (hunk 3 above)
Comment (davidnewhall) [reply Makefile]: ...and I just saw your other comment about how this works with go modules. I'm fine either way you want to swing this, but I'm kinda liking it _this_ way.

File: pkg/poller/prometheus.go [hunk 5]
```diff
@@ -24,6 +25,9 @@ func (u *UnifiPoller) RunPrometheus() error {
 		ReportErrors: true, // XXX: Does this need to be configurable?
 	}))
 
+	version.Version = Version
+	prometheus.MustRegister(version.NewCollector("unifipoller"))
```
Comment (davidnewhall) [inline pkg/poller/prometheus.go]: thinking about this too. I think, lets do `strings.Replace(appName, "-")` instead of `"unifipoller"`

Comment (SuperQ) [reply pkg/poller/prometheus.go]: I'd rather not, as there are a bunch of other illegal characters. Better is to use a defined `metricNamespace` that's shared with the Desc() calls.

File: Makefile (hunk 3 above)
Comment (SuperQ) [reply Makefile]: Yea, this is fine, it's easy to adjust the `-X` flag if the build process ever changes.

File: Makefile (hunk 4 above)
Comment (SuperQ) [reply Makefile]: I think it's fine this way for now.

**Comment by SuperQ** [comment]
Ok, I refactored the namespace value.

**Comment by davidnewhall** [comment]
Merging will also add new packages to the repo. hmu on discord and I'll give you the setup for that.

**Pull request merged**